vllm/v1/core/kv_cache_manager.py

```diff
@@ -79,30 +312,288 @@ def allocate_slots(
+
+    def _add_token_ids_to_blocks(self,
+                                 block_ids: List[int],
+                                 token_ids: List[int],
+                                 parent_block_id: Optional[int] = None) -> int:
+        """Add token IDs to a list of allocated blocks.
+        If a block becomes full after adding token IDs, cache it.
+        Return the token ID index that has not been added to the blocks
+        if the blocks are not enough to hold all the token IDs.
+
+        Args:
+            block_ids: A list of block IDs to add token IDs.
+            token_ids: A list of token IDs to add.
+            parent_block_id: The parent block ID. None if this is the
+                first block.
+
+        Returns:
+            The starting token ID index that has not been added to the blocks
+            due to insufficient given blocks.
+        """
+        parent_block = self.block_pool[
+            parent_block_id] if parent_block_id is not None else None
+        token_id_start = 0
+        for block_id in block_ids:
+            curr_block = self.block_pool[block_id]
+            curr_block.parent_block_id = parent_block_id
+
+            # If all token IDs are added, then the rest of the blocks are
+            # preallocated blocks, so we only need to update the
+            # parent_block_id.
+            if token_id_start == len(token_ids):
+                continue
+
+            # Add token IDs to the empty slots in the block.
+            empty_slots = self.block_size - len(curr_block.token_ids)
+            token_id_end = min(token_id_start + empty_slots, len(token_ids))
+            curr_block.token_ids.extend(token_ids[token_id_start:token_id_end])
+            # Cache the block if it becomes full.
+            if len(curr_block.token_ids) == self.block_size:
+                self._cache_full_block(curr_block, parent_block)
+            parent_block = curr_block
+            parent_block_id = parent_block.block_id
+            token_id_start = token_id_end
+        return token_id_start
+
+    def hash_prompt_tokens(self, token_ids: List[int]) -> List[int]:
+        """Computes hash values of a chain of blocks given a sequence of
+        token IDs. The hash value is used for prefix caching.
+
+        Args:
+            token_ids: A sequence of token ids in the prompt.
+
+        Returns:
+            The list of computed hash values.
+        """
+        ret = []
+        parent_block_hash = None
+        for start in range(0, len(token_ids), self.block_size):
+            end = start + self.block_size
+            block_token_ids = tuple(token_ids[start:end])
+            # Do not hash the block if it is not full.
+            if len(block_token_ids) < self.block_size:
+                break
+            block_hash = hash_block_tokens(parent_block_hash, block_token_ids)
+            ret.append(block_hash)
+            parent_block_hash = block_hash
+        return ret
+
+
+@lru_cache(maxsize=1024)
```

Comment:
Will `@lru_cache` itself involve hashing function parameters? I'm not sure whether it will actually accelerate this function.

Reply:
I have the same question too actually, so this may not really help...